k8sapiserver/k8sapiserver.go

```diff
@@ -81,199 +74,176 @@ func StartAPIServer(kubeAPIServerURL, etcdURL, frontendURL string) (*restclient.
 	}
 
-	// set the loopback client config
-	cfg.GenericConfig.LoopbackClientConfig = &restclient.Config{QPS: 50, Burst: 100, ContentConfig: restclient.ContentConfig{NegotiatedSerializer: legacyscheme.Codecs}}
-	cfg.GenericConfig.LoopbackClientConfig.Host = serverURL
-
-	privilegedLoopbackToken := uuid.New().String()
-	// wrap any available authorizer
-	tokens := make(map[string]*user.DefaultInfo)
-	tokens[privilegedLoopbackToken] = &user.DefaultInfo{
-		Name:   user.APIServerUser,
-		UID:    uuid.New().String(),
-		Groups: []string{user.SystemPrivilegedGroup},
+	if err = ioutil.WriteFile(saSigningKeyFile.Name(), []byte(ecdsaPrivateKey), 0o600); err != nil {
+		return nil, nil, xerrors.Errorf("write file %s failed: %v", saSigningKeyFile.Name(), err)
 	}
+	serverOpts.ServiceAccountSigningKeyFile = saSigningKeyFile.Name()
+	serverOpts.Authentication.ServiceAccounts.Issuers = []string{"https://foo.bar.example.com"}
+	serverOpts.Authentication.ServiceAccounts.KeyFiles = []string{saSigningKeyFile.Name()}
 
-	tokenAuthenticator := authenticatorfactory.NewFromTokens(tokens, cfg.GenericConfig.Authentication.APIAudiences)
-	cfg.GenericConfig.Authentication.Authenticator = authenticatorunion.New(tokenAuthenticator, authauthenticator.RequestFunc(alwaysEmpty))
-	tokenAuthorizer := authorizerfactory.NewPrivilegedGroups(user.SystemPrivilegedGroup)
-	cfg.GenericConfig.Authorization.Authorizer = authorizerunion.New(tokenAuthorizer, authorizerfactory.NewAlwaysAllowAuthorizer())
-
-	cfg.GenericConfig.LoopbackClientConfig.BearerToken = privilegedLoopbackToken
-
-	cfg.GenericConfig.PublicAddress = net.ParseIP("192.168.10.4")
-	cfg.GenericConfig.SecureServing = &genericapiserver.SecureServingInfo{Listener: fakeLocalhost443Listener{}}
-
-	cfg.GenericConfig.OpenAPIConfig = defaultOpenAPIConfig()
-
-	return cfg
-}
+	// disable admission plugins to avoid node taints, service account, cert approval etc.
+	serverOpts.Admission.GenericAdmission.DisablePlugins = apiserveropts.AllOrderedPlugins
+	serverOpts.Admission.GenericAdmission.EnablePlugins = []string{}
 
-type fakeLocalhost443Listener struct{}
+	// add cert directory to avoid permission error
+	certDir, err := os.MkdirTemp("", "apiserver-certs")
+	if err != nil {
+		return nil, nil, xerrors.Errorf("create temp certificate dir: %w", err)
+	}
+	serverOpts.SecureServing.ServerCert.CertDirectory = certDir
 
-func (fakeLocalhost443Listener) Accept() (net.Conn, error) {
-	return nil, nil
-}
+	cleanupFunc := func() {
+		os.RemoveAll(saSigningKeyFile.Name())
+		os.RemoveAll(certDir)
+	}
 
-func (fakeLocalhost443Listener) Close() error {
-	return nil
+	return serverOpts, cleanupFunc, nil
 }
 
-func (fakeLocalhost443Listener) Addr() net.Addr {
-	return &net.TCPAddr{
-		IP:   net.IPv4(127, 0, 0, 1),
-		Port: 443,
+func createK8SAPIChainedServer(etcdURL, frontendURL string) (*apiserver.APIAggregator, error) {
+	serverOpts, cleanupFunc, err := createK8SAPIServerOpts(etcdURL, frontendURL)
+	defer cleanupFunc()
```

Comment:
We shouldn't register `cleanupFunc to defer here, since cleanup will be executed when createK8SAPIChainedServer is finished.

Reply:
Make sense, I am going to pass the cleanupFunc all the way up to the shutdown Func instead.